tracer/src/Datadog.Trace/Ci/Agent/CIWriterHttpSender.cs

```diff
@@ -29,43 +30,98 @@ public CIWriterHttpSender(IApiRequestFactory apiRequestFactory)
             Log.Information("CIWriterHttpSender Initialized.");
         }
 
-        public async Task SendPayloadAsync(EventsPayload payload)
+        public async Task SendPayloadAsync(CIVisibilityProtocolPayload payload)
         {
-            var numberOfTraces = payload.Count;
-            var tracesEndpoint = payload.Url;
+            var payloadArray = payload.ToArray();
+            Log.Information<int, string>("Sending ({numberOfTraces} events) {bytesValue} bytes...", payload.Count, payloadArray.Length.ToString("N0"));
+            await SendPayloadAsync(
+                payload.Url,
+                static (request, payloadBytes) => request.PostAsync(new ArraySegment<byte>(payloadBytes), MimeTypes.MsgPack),
+                payloadArray).ConfigureAwait(false);
+        }
+
+        public async Task SendPayloadAsync(MultipartPayload payload)
+        {
+            Log.Information<int>("Sending {count} multipart items...", payload.Count);
+            await SendPayloadAsync(
+                payload.Url,
+                static (request, payloadArray) =>
+                {
+                    if (request is IMultipartApiRequest multipartRequest)
+                    {
+                        return multipartRequest.PostAsync(payloadArray);
+                    }
+
+                    ThrowHelper.ThrowInvalidOperationException("Sender doesn't support IMultipartApiRequest.");
```

Comment:
This obviously implies a config/programming error, but what happens if this throws here? Is it safe, or should we just log an error instead?

Reply:
I made a change to handle this scenario better in [08095f3](https://github.com/DataDog/dd-trace-dotnet/pull/2983/commits/08095f3ce582bf4337ed4e5e4187a9c50e531817)